h2/src/main/org/h2/mvstore/Page.java

```diff
@@ -260,50 +267,73 @@ static void readChildrenPositions(FileStore fileStore, long pos,
-                    chunkId, maxLength, pageLength);
+            throw DataUtils.newIllegalStateException(DataUtils.ERROR_FILE_CORRUPT,
+                    "File corrupted in chunk {0}, expected page length =< {1}, got {2}", chunkId, maxLength,
+                    pageLength);
         }
         buff.limit(start + pageLength);
         short check = buff.getShort();
         int m = DataUtils.readVarInt(buff);
         int mapId = collector.getMapId();
         if (m != mapId) {
-            throw DataUtils.newIllegalStateException(
-                    DataUtils.ERROR_FILE_CORRUPT,
-                    "File corrupted in chunk {0}, expected map id {1}, got {2}",
-                    chunkId, mapId, m);
+            throw DataUtils.newIllegalStateException(DataUtils.ERROR_FILE_CORRUPT,
+                    "File corrupted in chunk {0}, expected map id {1}, got {2}", chunkId, mapId, m);
         }
-        int checkTest = DataUtils.getCheckValue(chunkId)
-                ^ DataUtils.getCheckValue(offset)
+        int checkTest = DataUtils.getCheckValue(chunkId) ^ DataUtils.getCheckValue(offset)
                 ^ DataUtils.getCheckValue(pageLength);
         if (check != (short) checkTest) {
-            throw DataUtils.newIllegalStateException(
-                    DataUtils.ERROR_FILE_CORRUPT,
-                    "File corrupted in chunk {0}, expected check value {1}, got {2}",
-                    chunkId, checkTest, check);
+            throw DataUtils.newIllegalStateException(DataUtils.ERROR_FILE_CORRUPT,
+                    "File corrupted in chunk {0}, expected check value {1}, got {2}", chunkId, checkTest, check);
         }
         int len = DataUtils.readVarInt(buff);
         int type = buff.get();
         if ((type & 1) != DataUtils.PAGE_TYPE_NODE) {
-            throw DataUtils.newIllegalStateException(
-                    DataUtils.ERROR_FILE_CORRUPT,
+            throw DataUtils.newIllegalStateException(DataUtils.ERROR_FILE_CORRUPT,
                     "Position {0} expected to be a non-leaf", pos);
         }
+        /**
+         * The logic here is a little awkward. We want to (a) execute reads in parallel, but (b)
+         * limit the number of threads we create. This is complicated by (a) the algorithm is
+         * recursive and needs to wait for children before returning up the call-stack, (b) checking
+         * the size of the thread-pool is not reliable.
+         */
+        final List<Future<?>> futures = new ArrayList<>(len);
         for (int i = 0; i <= len; i++) {
-            collector.visit(buff.getLong());
+            final long childPagePos = buff.getLong();
+            if (executingThreadCounter.get() >= executorService.getMaximumPoolSize()) {
+                collector.visit(childPagePos, executorService, executingThreadCounter);
+            } else {
+                executingThreadCounter.incrementAndGet();
```

Comment:
It doesn't look like a really thread-safe implementation. `incrementAndGet()` may be greater than `executorService.getMaximumPoolSize()` here due to race between two threads in lines 312–315. I think that `compareAndSet()` should be used instead in a loop:
```Java
        for (int i = 0; i <= len; i++) {
            collector.visit(buff.getLong());
            final long childPagePos = buff.getLong();
            for (;;) {
                long counter = executingThreadCounter.get();
                if (counter >= executorService.getMaximumPoolSize()) {
                    collector.visit(childPagePos, executorService, executingThreadCounter);
                    break;
                } else {
                    if (executingThreadCounter.compareAndSet(counter, counter + 1)) {
                        Future<?> f = executorService.submit(new Runnable() {
                            @Override
                            public void run() {
                                try {
                                    collector.visit(childPagePos, executorService, executingThreadCounter);
                                } finally {
                                    executingThreadCounter.decrementAndGet();
                                }
                            }
                        });
                        futures.add(f);
                        break;
                    }
                }
            }
        }
```
(This modified code was not tested, it only illustrates the possible logic.)

Otherwise `executorService.submit()` may be rejected with an exception.

Reply:
thanks, pushed a fix